src/main/scala/is/hail/io/vcf/ExportVCF.scala

```diff
@@ -334,114 +292,125 @@ object ExportVCF {
       sb.append("#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO")
       if (hasSamples)
         sb.append("\tFORMAT")
-      vkds.sampleIds.foreach { id =>
+      vsm.sampleIds.foreach { id =>
         sb += '\t'
         sb.append(id)
       }
       sb.result()
     }
-
-    val idQuery: Option[Querier] = vas.getOption("rsid")
-      .filter {
-        case _: TString => true
-        case t => warn(
-          s"""found `rsid' field, but it was an unexpected type `$t'.  Emitting missing RSID.
-             |  Expected ${ TString() }""".stripMargin)
-          false
-      }.map(_ => vkds.queryVA("va.rsid")._2)
-
-    val qualQuery: Option[Querier] = vas.getOption("qual")
-      .filter {
-        case _: TFloat64 => true
-        case t => warn(
-          s"""found `qual' field, but it was an unexpected type `$t'.  Emitting missing QUAL.
-             |  Expected ${ TFloat64() }""".stripMargin)
-          false
-      }.map(_ => vkds.queryVA("va.qual")._2)
-
-    val filterQuery: Option[Querier] = vas.getOption("filters")
-      .filter {
-        case TSet(_: TString, _) => true
-        case t =>
-          warn(
-            s"""found `filters' field, but it was an unexpected type `$t'.  Emitting missing FILTERS.
-               |  Expected ${ TSet(TString()) }""".stripMargin)
-          false
-      }.map(_ => vkds.queryVA("va.filters")._2)
-
-    def appendRow(sb: StringBuilder, v: Variant, a: Annotation, gs: Iterable[T]) {
-
-      sb.append(v.contig)
-      sb += '\t'
-      sb.append(v.start)
-      sb += '\t'
-
-      sb.append(idQuery.flatMap(q => Option(q(a)))
-        .getOrElse("."))
-
-      sb += '\t'
-      sb.append(v.ref)
-      sb += '\t'
-      v.altAlleles.foreachBetween(aa =>
-        sb.append(aa.alt))(sb += ',')
-      sb += '\t'
-
-      sb.append(qualQuery.flatMap(q => Option(q(a)))
-        .map(_.asInstanceOf[Double].formatted("%.2f"))
-        .getOrElse("."))
-
-      sb += '\t'
-
-      filterQuery.flatMap(q => Option(q(a)))
-        .map(_.asInstanceOf[Set[String]]) match {
-        case Some(f) =>
-          if (f.nonEmpty)
-            f.foreachBetween(s => sb.append(s))(sb += ';')
-          else
-            sb.append("PASS")
-        case None => sb += '.'
+    
+    val fieldIdx = tva.fieldIdx
+    
+    def lookupVAField(fieldName: String, vcfColName: String, expectedTypeOpt: Option[Type]): (Boolean, Int) = {
+      fieldIdx.get(fieldName) match {
+        case Some(idx) =>
+          val t = tva.fields(idx).typ
+          if (expectedTypeOpt.forall(t == _)) // FIXME: make sure this is right
+            (true, idx)
+          else {
+            warn(s"export_vcf found va.$fieldName with type `$t', but expected type ${ expectedTypeOpt.get }. " +
+              s"Emitting missing $vcfColName.")
+            (false, 0)
+          }
+        case None => (false, 0)
       }
+    }
+    
+    val (idExists, idIdx) = lookupVAField("rsid", "ID", Some(TString()))
```

Comment:
Just a remark, I'll leave it up to you, but this might be cleaner if you pull the annotation manipulation code into a separate `VCFAnnotationView` class a la `HTSGenotypeView`.